### DISCUSSION TITLE
How to check for optional arguments of macro and modify expansion accordingly?

### DISCUSSION THREAD

**Discussion by stewmehr** [KaTeX/KaTeX#3520] [Q&A]
This is probably more of a `LaTeX` than a `KaTeX` question but maybe someone here can help me out.

In my *usual* `.tex` files I define personal macros with the `xparse` package. It should be possible to adapt most of them for use with `KaTeX` but I am struggling with the following one:

```latex
\DeclareDocumentCommand\setdef{mo}{\left\{#1\IfNoValueTF{#2}{}{ \mid #2}\right\}}
```
As you can see, the macro expansion changes depending on whether `IfNoValueTF{#2}` returns `true` or `false`. Is there a way to achieve something similar with `\def` or `\newcommand`, or some `KaTeX` trickery? I don't really mind if the solution requires a lot of low level `tex` code as long as I can automate parsing from `xparse` to `KaTeX`.

**Comment by edemaine** [comment]
You can use `\@ifnextchar[` (part of LaTeX, and how much of LaTeX's source code deals with optional arguments) combined with KaTeX's recently added support for `\def` patterns. For example:

```latex
\def\setdef#1{\left\{#1\@ifnextchar[\setdefMid\setdefEnd}
\def\setdefMid[#1]{\mid #1\setdefEnd}
\def\setdefEnd{\right\}}

\setdef{1,2,3};
\setdef{x^2}[x \in S]
```

![image](https://user-images.githubusercontent.com/2218736/151705961-b2d7dbf8-8e07-4937-bff5-e50565bf10f9.png)
